checks to see if membership editor is project and adds a project filter

Sean-McQ opened the pull request:
<!-- This template is for Devs to give QA details before moving the issue To-Test -->
### Summary
Fixes #8574 
Progresses #9341

This looks to be primarily a performance issue wherein loading a project edit page with a 'members' list (actually the `projectRoleTemplateBindings` resource) can either simply fail or get stuck in a repeating pattern at scale due to the way that `projectRoleTemplateBindings` scales across projects and clusters. Adding a project filter to the `projectRoleTemplateBindings` should cut down on the cost of this API call significantly.

### Occurred changes and/or fixed issues
The membership editor will now check if it's type is set to Norman's `projectRoleTemplateBindings` and will append a `projectId` field (Norman specific, Steve's `management.cattle.io.projectRoleTemplateBindings` uses `projectName` but is not used on this page.

While not strictly required by this PR, changes have been added to better support filtering and Sorting for Steve.

Changed out `uriEncodeComponent` for `uriEncode` since `uriEncodeComponent` was encoding characters that it shouldn't such as ":". Tested thoroughly and all existing tests still pass. New tests added to cover Steve specific url options.

### Technical notes summary
Please refer to the ReadMe on https://github.com/rancher/steve to verify URL options correctness for Steve. Specific Correctness of the "projectId" field for Norman's `projectRoleTemplateBindings` can be found on the collection itself under the `filters` key and the API call can also always be called directly or inspected in the network tab of the browser's developer tools to validate correctness. The actual spec used for Norman's query options can be found [here](https://github.com/rancher/api-spec/blob/master/specification.md#filtering).

### Areas or cases that should be tested
This is primarily a performance enhancement, if you'd like to actually see the specific performance impact, you'll want to create a rancher instance with a downstream cluster using ~400 projects and ~150 users on a single-node rancher instance but for the most part what is needed is just simple functional validation. The members list does load, it shows the correct members, no broken API calls, etc...

### Areas which could experience regressions
No Steve API calls were using filters and/or sorting in them so nothing should break there, all existing unit tests on urlOptions still pass so I'm not assuming any regressions there either.

### Screenshot/Video
<!-- Attach screenshot or video of the changes and eventual comparison if you find it necessary -->

aalves08 reviewed (commented):
@Sean-McQ changes seem to be "trivial" but I don't have any info on how to reproduce this and also I don't know about this `filter` option on the `findAll` 😛 . I would say that if @mantis-toboggan-md is okay with this change, I'm also okay with it.

Sean-McQ commented:
Changed out uriEncodeComponent for uriEncode since uriEncodeComponent was encoding characters that it shouldn't such as ":". Tested thoroughly and all existing tests still pass. While I was in there I also addressed the fact that filters and sorting criteria for Steve requests wouldn't be formatted correctly for that API so I added in some code to support that. New tests added to cover Steve specific url options. There's still quite a bit of functionality we're not using in Steve's URL options but that's covered under #9341

aalves08 commented on shell/components/form/Members/MembershipEditor.vue:
```diff
@@ -56,8 +56,13 @@ export default {
+    const roleBindingRequestParams = { type: this.type, opt: { force: true } };
+
+    if (this.type === NORMAN.PROJECT_ROLE_TEMPLATE_BINDING) {
+      Object.assign(roleBindingRequestParams, { opt: { filter: { projectId: this.parentId.split('/').join(':') } } });
```
Is there any chance of `this.parentId` might be `undefined`? If so then `this.parentId.split` would throw an error. (optional chaining)

Sean-McQ replied:
It shouldn't be possible for it to be undefined or null while the type is `NORMAN.PROJECT_ROLE_TEMPLATE_BINDING` since the route requires it but it's trivial to add code to ensure that that's so prior to running so I'll do that.

aalves08 commented on shell/plugins/steve/__tests__/getters.spec.ts:
```diff
@@ -86,8 +92,17 @@ describe('steve: getters', () => {
+    });
+    it('returns a string with a sorting criteria formatted for steve if the sort option is provided and an order if sortOrder is "desc" and the url starts with "/v1"', () => {
+      expect(urlOptionsGetter('/v1/foo', { sortBy: 'bar', sortOrder: 'desc' })).toBe('/v1/foo?exclude=metadata.managedFields&sort=-bar');
+    });
```
Very interesting all of these options! never knew about these. Nice! 🤜

Sean-McQ replied:
Thank @cmurphy, she gave us all sorts of goodies in Steve!

aalves08 reviewed (commented):
just a small comment, but otherwise looks good. Still need to test it. Once you have that system ready, let me know 🙏

aalves08 commented:
All good @Sean-McQ ! 🚢 it

aalves08 reviewed (approved):
LGTM!
